### DISCUSSION TITLE
Update Entity away from TableController

### DISCUSSION THREAD

**Discussion by henda79** [Azure/azure-mobile-apps#676] [Q&A] [answered]
I have an API which uploads a file, when the file is uploaded I need to update a column, but if I do this without updating `Version` and `UpdateAt` I get a precondition error on next sync.

What is the best way to go about this ?

**Comment by adrianhall** [comment, marked as answer]
You can make an instance of the repository that does the column update.  The repository is ephemeral in most cases (certainly, EF is ephemeral - InMemory is the only one that isn't thus far).  Goign through the repository will ensure the version and updatedAt are updated at the same time.

**Reply by henda79** [reply]
Thanks that seems to have worked.

Am I Ok just to new up an instance of EntityTableRepository in another controller ?

**Reply by adrianhall** [reply]
Yes - EntityTableRepository is stateless - all state is managed via the EF Core DbContext - so you can spin one up with a new DbContext at any point.

**Reply by ideas2appsaustralia** [reply]
Dears
I need to add and update to the SQL Db that I use for a Mobile App based on DataSyncCLient and its API using EntityTableData class. Im building a Blazor WASM website to use for administration of the data. For Add, update etc I need to populate "Version" column.  Could I set the property like below for the DbContext which can automatically handle it leavig me to include code to compare versions ?

modelBuilder.Entity("tablename", b =>
            {
                b.Property<byte[]>("Timestamp")
                    .IsConcurrencyToken()
                    .ValueGeneratedOnAddOrUpdate();
            });

**Reply by adrianhall** [reply]
You don't need to.  The current EF Core version of EntityTableData and EntityTableRepository do that for you.  This is done via attribute in the EntityTableData.

What you DO need to do is ensure you update the UpdatedAt field to the date/time of the last update.  Just set the field to DateTimeOffset.UtcNow whenever you update the record.  EntityTableRepository does this for you, but if you are modifying the DbContext outside of the repository, you need to do it.

Note that I've found the ValueGeneratedOnAddOrUpdate() to be unreliable across databases when it comes to DateTimeOffset.  They even have a section on this: https://learn.microsoft.com/en-us/ef/core/modeling/generated-properties?tabs=fluent-api#update-timestamp - since we work with multiple databases, it's not reasonable for the default repository implementation to update the UpdatedAt.  However, if you are explicitly working with SQL Server (or Azure SQL) and want to do the trigger instead (so your web front end doesn't have to worry about it), then you need to remove the UpdatedAt updates from the EntityTableRepository.  Just create a copy of the code and remove the un-necessary bits, then use your new and improved version as your repository in your table controllers.

**Reply by ideas2appsaustralia** [reply]
Adrian, bit more help if you dont mind

EntityTableData is not available in my blazor WASM server namespace. Do I bring it in with just "using Microsoft.AspNetCore.Datasync.EFCore;" ?

I"ll try it anyway

scott

**Reply by adrianhall** [reply]
You wouldn't use EntityTableData in your client application.  You can just copy over the DatasyncClientData class (it's small and is just there to provide a reference for the properties needed in your own entities), or you can just include the definitions of the properties that are in DatasyncClientData in your own entieies.
